MM-48103: Adds request for persistent notification

koox00 opened the pull request:
#### Summary

- Adds the UI for requesting persistent notifications when selecting a post's priority.
- Dispatches a desktop notification when `persistent_notification_triggered` websocket events is handled. This particular notification bypasses user's channel notify preferences.
- Adds admin console settings for persistent notifications.
- Adds a prop `shouldDisable` to the `AdvancedTextEditor` so we can control `disable` from the outside world.
- Adds persistent notifications specific validation before sending a message with `persistent_notifications: true`. 1. Users must be mentioned, 2. you cannot have channel wide mentions, and 3. there is a limit to how many users you can mention. 
- Adds a confirmation modal to be displayed and inform the user posting that the message will trigger `persistent_notifications` to users mentioned.
- Refactors priority specific code outside of the `AdvnacedCreatePost` and into separate components.
- Refactors `PostPriorityPicker` to use `floating-ui` instead of bootstrap popover, and tooltip. 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-48103
https://mattermost.atlassian.net/browse/MM-48689

#### Related Pull Requests

- Has server changes (https://github.com/mattermost/mattermost-server/pull/21619)

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added option to request persistent notifications upon creating urgent posts
Added admin console settings for persistent notifications
Triggers a desktop notification for persistent notifications
```

koox00 committed:
MM-48103: Adds request for persistent notification

Adds the UI for requesting persistent notifications when selecting a
post's priority

koox00 committed:
Adds intl

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Adds confirmation modal, and validation

- Disable the send button when there are no mentions and persistent
  notifications are requested
- Show a confirmation modal when the user tries to send a message with
  persistent notifications
- Forbid the user from posting if the number of mentions is more than 5

koox00 committed:
Adds admin console settings for persistent notifications

koox00 committed:
Adds intl

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Various fixes

koox00 committed:
Fixes tests

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Gets the correct number of mentions in the message

koox00 committed:
Adds persistent notifications in drafts UI

koox00 committed:
Adds no recipients clause in confirmation modal

koox00 committed:
Guests should not be allowed by default

koox00 committed:
Satisfies linters

koox00 committed:
Fixes text

koox00 committed:
Fixes snapshot tests

koox00 committed:
Fixes isPersistentNotificationsEnabled

koox00 committed:
Handle persistent desktop notifications

When a persistent notification is triggered
(ws 'persistent_notification_triggered') we need to send a desktop
notification, which overrides channel members notify props, to all
mentioned users.

This commit handles that case by bypassing some checks when
'sendDesktopNotification' receives the 'persistent' argument as true.
Which is only set to true when a 'persistent_notification_triggered'
event is received.

koox00 committed:
Handles @channel, @all, @here

Channel wide mentions are not allowed when requesting persistent
notifications

koox00 committed:
Adds intl

koox00 committed:
Fixes types

koox00 committed:
Fixes tests

koox00 committed:
Fixes tests

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Addresses review comments, and refactors

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Fixes toggle description text to include interval

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
MM-49212,MM-49205: fixes issues found in testing

When the confirmation modal is open. Clicking ENTER should accept and send the message.
In GMs, messages should not be sent unless user(s) are mentioned.

koox00 commented:
/update-branch

mattermod committed:
Merge branch 'master' into persist-notifications

vish9812 committed:
Merge remote-tracking branch 'origin/master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

hmhealey commented:
Sorry for the delays on reviewing this. I'll make sure to take a look at this tomorrow

michelengelen commented on components/drafts/channel_draft/channel_draft.tsx:
```diff
@@ -65,11 +93,16 @@ function ChannelDraft({
-        history.push(channelUrl);
-    }, [value, channelUrl, user.id, channel.id]);
+        if (
+            postPriorityEnabled &&
+            value?.metadata?.priority?.priority === PostPriority.URGENT &&
+            value?.metadata?.priority?.persistent_notifications
```
nit: this is the second time this check appears, so we could create a helper for that

michelengelen commented on components/drafts/channel_draft/index.ts:
```diff
@@ -24,10 +25,12 @@ function makeMapStateToProps() {
 
         const teamId = getCurrentTeamId(state);
         const channelUrl = getChannelURL(state, channel, teamId);
+        const postPriorityEnabled = isPostPriorityEnabled(state);
```
```suggestion
```

michelengelen commented on components/drafts/channel_draft/index.ts:
```diff
@@ -24,10 +25,12 @@ function makeMapStateToProps() {
         return {
             channel,
             channelUrl,
+            postPriorityEnabled,
```
nit:
```suggestion
            postPriorityEnabled: isPostPriorityEnabled(state),
```

michelengelen reviewed (approved):
This is looking great ... just a few small nits!

hmhealey commented on actions/notification_actions.jsx:
```diff
@@ -68,17 +68,21 @@ export function sendDesktopNotification(post, msgProps) {
         const isCrtReply = isCollapsedThreadsEnabled(state) && post.root_id !== '';
 
-        if (!member || isChannelMuted(member) || userStatus === UserStatuses.DND || userStatus === UserStatuses.OUT_OF_OFFICE) {
+        if (!member || (!persistent && isChannelMuted(member)) || userStatus === UserStatuses.DND || userStatus === UserStatuses.OUT_OF_OFFICE) {
```
0/5 This is a comment about the design, but having these bypass mute feels rather obnoxious. Generally, you wouldn't have a channel muted if it would contain notifications like this anyway

koox00 replied:
cc @esethna

hmhealey commented on actions/notification_actions.jsx:
```diff
@@ -68,17 +68,21 @@ export function sendDesktopNotification(post, msgProps) {
 
-        let notifyLevel = member?.notify_props?.desktop || NotificationLevels.DEFAULT;
+        // we are bypassing channel member's notify props in the case of persistent notifications
+        let notifyLevel = (!persistent && member?.notify_props?.desktop) || NotificationLevels.DEFAULT;
```
0/5 Similarly, bypassing the desktop notification setting seems odd. If a user really wanted to turn off desktop notifications, they could still get around it by disabling them at an OS level

koox00 replied:
It's only bypassing the channel specific `notify_prop` for desktop notifications.
User `notify_props` are respected.

cc @esethna

hmhealey commented on actions/notification_actions.jsx:
```diff
@@ -68,17 +68,21 @@ export function sendDesktopNotification(post, msgProps) {
         }
 
-        if (notifyLevel === NotificationLevels.NONE) {
+        if (persistent && mentions.indexOf(currentUserId) === -1) {
```
3/5 This seems wrong to me. I would think that only people who are supposed to get a persistent notification would receive a websocket event for it.

koox00 replied:
We could change it to send only to mentioned users, that's true.

cc @vish9812

koox00 replied:
Yes we are changing this before merging, Vishal is working on it on the server side

hmhealey commented on components/admin_console/admin_definition.jsx:
```diff
@@ -2826,6 +2826,136 @@ const AdminDefinition = {
+                        help_text_values: {
+                            link: (msg) => (
+                                <a
+                                    href='https://mattermost.com/pl/message-priority/'
```
I think the `pl` here stands for permalink rather than the language

michelengelen replied:
ah, ok ... good to know

hmhealey commented on components/advanced_create_post/advanced_create_post.tsx:
```diff
@@ -1519,7 +1525,38 @@ class AdvancedCreatePost extends React.PureComponent<Props, State> {
+            return false;
+        }
+
+        const mentions = mentionsMinusSpecialMentionsInText(this.state.message);
```
Note that this function just counts things that look like mentions, so this will allow someone to enable persistent notifications for a message that just contains a fake mention for a user that doesn't exist. It also seems to capture mentions in things like code blocks that otherwise wouldn't trigger mentions

koox00 replied:
This is only for a quick validation I didn't want to burden the `onChange` much.
We are validating usernames in the confirmation modal.

I didn't think of code blocks at all though, let me check on that.

hmhealey commented on components/post_priority/post_priority_picker_overlay.tsx:
```diff
@@ -1,56 +1,146 @@
+                                position: pickerStrategy,
+                                top: pickerY ?? 0,
+                                left: pickerX ?? 0,
+                                zIndex: 999,
```
0/5 - Needing to set a z-index here seems weird since this is being portalled to the root level already. I would think that it's above the rest of the DOM simply because of its position in it

hmhealey commented on entry.tsx:
```diff
@@ -28,7 +28,10 @@ declare global {
-        if (msg === 'ResizeObserver loop limit exceeded') {
+        if (
+            msg === 'ResizeObserver loop limit exceeded' ||
+            msg === 'ResizeObserver loop completed with undelivered notifications.'
```
What is this being added for? This error message seems to indicate an infinite loop which seems really concerning

koox00 replied:
@hmhealey the issue was happening when we select "urgent" in the picker, which adds a new toogle in the ui. Floating-ui has to watch for element resizes to recompute the position.

According to this issue https://github.com/floating-ui/floating-ui/issues/1740#issuecomment-1151773854, it was safe to ignore.

I had tried the proposed solution `witdh: max-content` on the floating element, but that didn't solve the issue.

However, after merging latest master I can no longer reproduce :/ even with max-content removed :/ :/ ...
I'm removing the `msg === 'ResizeObserver loop completed with undelivered notifications.'`.

NOTE that the first condition to ignore was there beforehand, I have no idea why though.

koox00 commented on components/advanced_text_editor/advanced_text_editor.tsx:
```diff
@@ -101,6 +101,7 @@ type Props = {
     isThreadView?: boolean;
     additionalControls?: React.ReactNodeArray;
     labels?: React.ReactNode;
+    shouldDisable?: boolean;
```
`disabled` was used as internal state, I'll rename to `disableSend`

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 commented on components/advanced_create_post/advanced_create_post.tsx:
```diff
@@ -882,7 +899,7 @@ class AdvancedCreatePost extends React.PureComponent<Props, State> {
     }
 
-    handleDraftChange = (draft: PostDraft) => {
+    handleDraftChange = (draft: PostDraft, instant = false) => {
```
We have a `setTimeout` below, but this also makes selecting a post priority to be saved after 500ms in the draft, but we want it to happen instantly.

koox00 replied:
@michelengelen ☝️

koox00 committed:
Changes according to reviews

koox00 committed:
Fixes that mentions in code blocks are not counted

koox00 committed:
Fixes linter

koox00 committed:
Fixes snapshots

koox00 commented:
/check-cla

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Reverts notification bypasses

koox00 commented:
/update-branch

mattermost-build committed:
Merge branch 'master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Validates persistent notifications interval setting

Interval cannot be less than 2 minutes.

koox00 committed:
Changes validation text to match the API

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

hmhealey reviewed (approved):
Looks like the changes to only send websocket events to affected users went in, so if that's the case, this looks good to me

hmhealey commented:
@koox00 This may need the server PR to get merged first depending on how finnicky the test servers are today, but perhaps this should get a QA review before it's merged?

koox00 commented:
> Looks like the changes to only send websocket events to affected users went in, so if that's the case, this looks good to me

@hmhealey yep we passed your comments to product and decided that it's better without overriding! In the future depending on user feedback this might change though.

> @koox00 This may need the server PR to get merged first depending on how finnicky the test servers are today, but perhaps this should get a QA review before it's merged?

QA/UX reviews are will be both in the server since we already have comments there.
We will merge at the same time but I don't expect that to happen today.

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Fixes initial focus

Initial focus was focusing on the feedback link,
this commit removes initial focus, so you'll need to tab to change
focus.

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Fixes z-index in <768, and error messages

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Fixes CI build

koox00 committed:
Fixes tests

michelengelen commented on components/advanced_create_post/advanced_create_post.tsx:
```diff
@@ -1553,9 +1553,12 @@ class AdvancedCreatePost extends React.PureComponent<Props, State> {
         return mentions.length > 0;
     }
 
+    getSpecialMentions = (): {[key: string]: boolean} => {
+        return specialMentionsInText(this.state.message);
+    }
```
nit: 
```suggestion
    getSpecialMentions = (): {[key: string]: boolean} => specialMentionsInText(this.state.message);
```

michelengelen commented on components/persist_notification_confirm_modal.tsx:
```diff
@@ -67,12 +69,9 @@ function PersistNotificationConfirmModal({
-                id='persist_notification.special_mentions.description'
-                defaultMessage='Cannot use @channel, @all or @here to mention recipients of persistent notifications'
-            />
+            <HasSpecialMentions specialMentions={specialMentions}/>
```
nit: you can also put this on one line without parantheses

michelengelen commented on components/post_priority/error_messages.tsx:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+export function HasSpecialMentions({specialMentions}: {specialMentions: {[key: string]: boolean}}) {
+    return (
+        <FormattedMessage
+            id={'post_priority.error.special_mentions'}
+            defaultMessage={'Cannot use @{mention} to mention recipients of persistent notifications'}
+            values={{
+                mention: getFirstSpecialMention(specialMentions),
+            }}
+        />
+    );
+}
+
+export function HasNoMentions() {
+    return (
+        <FormattedMessage
+            id={'post_priority.error.no_mentions'}
+            defaultMessage={'Recipients must be @mentioned'}
+        />
+    );
+}
```
just and idea: Can't we make this one component?

Something along this line:
```
function getFirstSpecialMention(specialMentions: {[key: string]: boolean}): string {
    switch (true) {
    case specialMentions.all:
        return 'all';
    case specialMentions.here:
        return 'here';
    case specialMentions.channel:
        return 'channel';
    default:
        return '';
    }
}

export function SpecialMentions({specialMentions}: {specialMentions: {[key: string]: boolean}}) {
    const firstSpecialMention = getFirstSpecialMention(specialMentions);
    return firstSpecialMention ? (
        <FormattedMessage
            id={'post_priority.error.special_mentions'}
            defaultMessage={'Cannot use @{mention} to mention recipients of persistent notifications'}
            values={{
                mention: firstSpecialMention,
            }}
        />
    ) : (
        <FormattedMessage
            id={'post_priority.error.no_mentions'}
            defaultMessage={'Recipients must be @mentioned'}
        />
    );
}
```

Thoughts?

michelengelen reviewed (approved):
just a few nits and an idea that could maybe improve this a bit ... but overall looks really solid! Nice work!

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Changes error message for special mentions

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

vish9812 committed:
rename config

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Merge remote-tracking branch 'upstream/master' into persist-notifications

koox00 committed:
Shrug

lieut-data commented:
Heads up that as part of our efforts to move to a monorepo, we're closing out a number of older pull requests like this one to help streamline the effort.

If you'd like to preserve these changes -- even if you're not the original author! -- feel free to resubmit the pull request against the monorepo once it's ready. You can subscribe to [mattermost-server-issue-22420](https://www.github.com/mattermost/mattermost-server/issues/22420) for status updates on this effort.

koox00 commented:
I've added this PR in the server PR ﻿﻿mattermost/mattermost-server#21619
cc @hmhealey
